### DISCUSSION TITLE
[Berry] Read file line by line without Tasmota locking up

### DISCUSSION THREAD

**Discussion by Slyke** [arendst/Tasmota#15806] [Q&A] [answered]
Hello, trying to figure out how to process a file line by line. I have a large JSON file, it's basically an array of objects like:
```
[
  {"hasBeenUploaded": true, "someData": "aboutSomeStuff1"},
  {"hasBeenUploaded": true, "someData": "aboutSomeStuff2"},
  {"hasBeenUploaded": true, "someData": "aboutSomeStuff3"},
  {"hasBeenUploaded": false, "someData": "aboutSomeStuff4"},
  {"hasBeenUploaded": false, "someData": "aboutSomeStuff5"},
  {"hasBeenUploaded": false, "someData": "aboutSomeStuff6"},
  {"hasBeenUploaded": false, "someData": "aboutSomeStuff7"}
]
```

I want to loop through each line of the file, upload the JSON object to the API server, then update the line's `hasBeenUploaded` property to true, save the file and move onto the next line.

Right now I'm figuring out the loop part, and have the following code:
```
  import path
  var logFile = open('/sd/sensor_log.json', 'r')
  var lineBuf = " "
  while (lineBuf != nil && lineBuf != "") # I would expect readline() to return nil at EoF, but it apparently just returns an empty string. Either way I check for both.
    tasmota.yield()
    tasmota.delay(5)
    lineBuf = logFile.readline()
    tasmota.yield()
    print("lineBuf", lineBuf)
  end
```

When running this code, it causes the Tasmota device to completely lock up and requires a hardware reset or wait for it to complete the loop. I have tried both yield and delay independently, but left them both in my code example for brevity.

**Comment by sfromis** [comment, marked as answer]
Using `delay` helps nothing, but makes it worse, as it is still blocking code. Using `yield` is more to feed the watchdog, and allow low level functions to kick in, not so much to have Tasmota being responsive while in a long time loop. 

You might be able to have your cake and eat it too by using a `tasmota.set_timer` function to read chunks of the file, allowing each timer task to exit while the file is still open.

In general, avoid using loops in Berry code under Tasmota to manage progress.

You should have paid close attention to this part of the docs:
> Berry code, when it is running, blocks the rest of Tasmota. This means that you should not block for too long, or you may encounter problems. As a rule of thumb, try to never block more than 50ms.

**Reply by Slyke** [reply]
The best way I came up with solving this is to write a driver and put the file readline on every second. It checks if there's already a request in progress before attempting to read. This spaces out the read and uploading enough to not lock up the system, but still upload at a reasonable rate.

**Reply by sfromis** [reply]
I'd suggest a simpler setup than using a driver and callbacks like every second.

You can have a function reading as much as is workable in one go (one or more lines), and then have it do a `tasmota.set_timer` call to schedule itself after some interval, which can be as low as zero. The point is that the scheduling will occur in the next main loop tick after Tasmota has had the chance to do other work. No need to check for what already may be in progress, and you can run near full speed instead of unwanted delays.
